search/searcher/search_conjunction.go

```diff
@@ -205,12 +220,45 @@ OUTER:
 			// don't bump i, so that we'll examine the just-advanced
 			// currs[i] again
 		}
+		// if we get here, a doc matched all readers
+		if s.searcherPositions == nil {
+			rv = s.scorer.Score(ctx, s.currs)
+		} else {
+			// we only score the document if the relative positioning rule is satisfied
+			// if not, we advance the maxIDIdx and try again
+			var hitTable = make([][]PositionPair, len(s.currs))
+			var hitValidityTable = make([][]bool, len(s.currs))
+			for id, match := range s.currs {
+				hitTable[id] = make([]PositionPair, len(match.FieldTermLocations))
+				firstPos := match.FieldTermLocations[0].Location.Pos
+				lastPos := firstPos
+				colID := 0
+				for i := 1; i < len(match.FieldTermLocations); i++ {
+					if match.FieldTermLocations[i].QueryIndex > match.FieldTermLocations[i-1].QueryIndex ||
+						match.FieldTermLocations[i].OccurrenceIndex > match.FieldTermLocations[i-1].OccurrenceIndex {
+						hitTable[id][colID] = PositionPair{firstPos, lastPos}
+						colID++
+						firstPos = match.FieldTermLocations[i].Location.Pos
```

Comment:
trying to understand the purpose of these checks over here, the way i see it queryIndex is like specific to a documentMatch and the occurrenceIndex is specifc to a location within a doc (and the doc can have multiple locations for the term) (understood this bit by seeing the MergeFieldTermLocations() API whose result is ). the structure of the slice is such that there are multiple location values and they are grouped by docMatches (or the queryIndex). i'm not getting in what case would the the occurrenceIndex be same, other than across queryIndex but we are anyways checking the adjacent ones (whose occurenceIndex wouldn't be same within a queryindex group). also, why should the firstPos be the same for FTLs across doc matches (for example if the queryIndex 0 and 1 had a single location value each, then the firstPos would be the same but the lastPos be changed. is that expected behaviour? does this snippet work as expected when there is a highly nested conjunction search? (conjunction within a conjunction within a conjunction etc) because in that case the structure of the FieldTermLocations wouldn't be trivial like a single level nested FTL)